hive-metastore/src/main/java/org/apache/iceberg/hive/MetastoreLock.java

```diff
@@ -0,0 +1,533 @@
+        // the maximum number of
+        // attempts the Tasks.run() would try is `retries + 1`. Here, for checking locks, we use
+        // timeout as the
+        // upper bound of retries. So it is just reasonable to set a large retry count. However, if
+        // we set
+        // Integer.MAX_VALUE, the above logic of `retries + 1` would overflow into
+        // Integer.MIN_VALUE. Hence,
+        // the retry is set conservatively as `Integer.MAX_VALUE - 100` so it doesn't hit any
+        // boundary issues.
+        Tasks.foreach(lockInfo.lockId)
+            .retry(Integer.MAX_VALUE - 100)
+            .exponentialBackoff(lockCheckMinWaitTime, lockCheckMaxWaitTime, lockAcquireTimeout, 1.5)
+            .throwFailureWhenFinished()
+            .onlyRetryOn(WaitingForLockException.class)
+            .run(
+                id -> {
+                  try {
+                    LockResponse response = metaClients.run(client -> client.checkLock(id));
+                    LockState newState = response.getState();
+                    lockInfo.lockState = newState;
+                    if (newState.equals(LockState.WAITING)) {
+                      throw new WaitingForLockException(
+                          String.format(
+                              "Waiting for lock on table %s.%s", databaseName, tableName));
+                    }
+                  } catch (InterruptedException e) {
+                    Thread.interrupted(); // Clear the interrupt status flag
+                    LOG.warn(
+                        "Interrupted while waiting for lock on table {}.{}",
+                        databaseName,
+                        tableName,
+                        e);
+                  }
+                },
+                TException.class);
+      }
+    } catch (WaitingForLockException e) {
+      timeout = true;
+      duration = System.currentTimeMillis() - start;
+    } catch (TException e) {
+      error = e;
+    } finally {
+      if (!lockInfo.lockState.equals(LockState.ACQUIRED)) {
+        unlock(Optional.of(lockInfo.lockId));
+      }
+    }
+
+    if (lockInfo.lockState.equals(LockState.ACQUIRED)) {
+      return lockInfo.lockId;
+    }
+
+    // timeout and do not have lock acquired
+    if (timeout) {
+      throw new LockException(
+          "Timed out after %s ms waiting for lock on %s.%s", duration, databaseName, tableName);
+    }
+
+    // On thrift error and do not have lock acquired
+    if (error != null) {
+      throw new LockException(
+          error, "Metastore operation failed for %s.%s", databaseName, tableName);
+    }
+
+    // Just for safety. We should not get here.
+    throw new LockException(
+        "Could not acquire the lock on %s.%s, lock request ended in state %s",
+        databaseName, tableName, lockInfo.lockState);
+  }
+
+  /**
+   * Tries to create a lock. If the lock creation fails, and it is possible then retries the lock
```

Comment:
Nit: "Creates a lock, retrying if possible on failure"?  More concise.

Also I think we can remove "If the lock creation is successful then a {@link LockInfo} is returned,
   * otherwise an appropriate exception is thrown.", and move corresponding information to return , and throws elements, which are already in the comment.

Reply:
Thanks for the comment - did the appropriate changes in #6648. Please check there